crates/swc_ecma_lints/src/rules/valid_typeof.rs

```diff
@@ -0,0 +1,139 @@
+use serde::{Deserialize, Serialize};
+use swc_common::{errors::HANDLER, Span};
+use swc_ecma_ast::*;
+use swc_ecma_visit::{noop_visit_type, Visit, VisitWith};
+
+use crate::{
+    config::{LintRuleReaction, RuleConfig},
+    rule::{visitor_rule, Rule},
+};
+
+const MESSAGE: &str = "Invalid typeof comparison value";
+
+const VALID_TYPES: &[&str] = &[
+    "undefined",
+    "object",
+    "boolean",
+    "number",
+    "string",
+    "function",
+    "symbol",
+    "bigint",
+];
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ValidTypeofConfig {
+    require_string_literals: Option<bool>,
+}
+
+pub fn valid_typeof(config: &RuleConfig<ValidTypeofConfig>) -> Option<Box<dyn Rule>> {
+    match config.get_rule_reaction() {
+        LintRuleReaction::Off => None,
+        _ => Some(visitor_rule(ValidTypeof::new(config))),
+    }
+}
+
+#[derive(Debug, Default)]
+struct ValidTypeof {
+    expected_reaction: LintRuleReaction,
+    require_string_literals: bool,
+}
+
+impl ValidTypeof {
+    fn new(config: &RuleConfig<ValidTypeofConfig>) -> Self {
+        let rule_config = config.get_rule_config();
+
+        Self {
+            expected_reaction: config.get_rule_reaction(),
+            require_string_literals: rule_config.require_string_literals.unwrap_or(true),
+        }
+    }
+
+    fn emit_report(&self, span: Span) {
+        HANDLER.with(|handler| match self.expected_reaction {
+            LintRuleReaction::Error => {
+                handler.struct_span_err(span, MESSAGE).emit();
+            }
+            LintRuleReaction::Warning => {
+                handler.struct_span_warn(span, MESSAGE).emit();
+            }
+            _ => {}
+        });
+    }
+
+    fn check(&self, span: Span, str_operand: &str) {
+        if !VALID_TYPES.contains(&str_operand) {
+            self.emit_report(span);
+        }
+    }
+}
+
+impl Visit for ValidTypeof {
+    noop_visit_type!();
+
+    fn visit_bin_expr(&mut self, bin_expr: &BinExpr) {
+        if let op!("==") | op!("===") | op!("!=") | op!("!==") = bin_expr.op {
+            match (bin_expr.left.as_ref(), bin_expr.right.as_ref()) {
+                // case typeof x === "type"
+                (
+                    Expr::Unary(UnaryExpr {
+                        op: UnaryOp::TypeOf,
```

Comment:
Can you change this to `op: op!("typeof")`?

Reply:
Forget about it every time 😅